### PR TITLE
refactor(docker): Use pre/post commands instead of wrapping

### PR DESCRIPTION
**Pull request by zharinov**
ref #2863

### CODE REVIEW

File: test/util/exec.spec.ts [hunk 1]
```diff
@@ -7,71 +7,115 @@ import { exec, ExecOptions } from '../../lib/util/exec';
...
+      { docker: { image, tag } },
+      `docker run --rm ${image}:${tag} ${cmd}`,
+      { encoding },
+    ],
+
+    // CWD mounting
+    [
+      cmd,
+      { docker: { image }, cwd },
+      `docker run --rm -v "${cwd}":"${cwd}" -w "${cwd}" ${image} ${cmd}`,
```
Comment (zharinov) [inline test/util/exec.spec.ts]: When passing `cwd`, we always have to mount it too, aren't we?

File: lib/util/docker/index.ts [hunk 2]
```diff
@@ -7,37 +9,53 @@ export interface DockerOptions {
...
+    preCommands = [],
+    postCommands = [],
+  } = options;
 
   const result = ['docker run --rm'];
   if (dockerUser) result.push(`--user=${dockerUser}`);
 
-  if (volumes)
-    result.push(...volumes.filter(x => !!x).map(vol => `-v "${vol}":"${vol}"`));
+  let volumes = options.volumes || [];
```
Comment (ViceIce) [inline lib/util/docker/index.ts]:
```suggestion
  let volumes = new Set(options.volumes || []);
```

File: lib/util/docker/index.ts [hunk 3]
```diff
@@ -7,37 +9,53 @@ export interface DockerOptions {
...
+  volumes = cwd ? [...volumes, options.cwd] : volumes;
+  if (volumes) {
+    result.push(
+      ...uniq(volumes)
```
Comment (ViceIce) [inline lib/util/docker/index.ts]:
```suggestion
      ...Array.from(volumes)
```

File: lib/util/docker/index.ts [hunk 4]
```diff
@@ -7,37 +9,53 @@ export interface DockerOptions {
...
-      );
-    }
-  } else {
-    result.push(cmd);
-  }
+  const commands = [...preCommands, cmd, ...postCommands];
+  const command =
+    commands.length === 1
+      ? cmd
+      : `bash -l -c "${commands.join(' && ').replace(/"/g, '\\"')}"`;
```
Comment (ViceIce) [inline lib/util/docker/index.ts]: Do we have proper escaping for all cases here?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/docker/index.ts (hunk 2 above)
Comment (ViceIce) [reply lib/util/docker/index.ts]: not required

File: lib/util/docker/index.ts (hunk 3 above)
Comment (ViceIce) [reply lib/util/docker/index.ts]: ok to use `uniq`

File: lib/util/docker/index.ts (hunk 4 above)
Comment (ViceIce) [reply lib/util/docker/index.ts]: We have to make sure that escaping quotes works for complicated calls.

**Review by viceice — APPROVED** [review]

**Pull request merged**